_posts/2017-09-24-Unit-testing-in-angular.md

```diff
@@ -0,0 +1,504 @@
+If there's any subcomponent in the HTML of the component you're testing, they should either be imported through a module or added in the declarations in the TestBed. 
+
+If the Angular router is used in the view of your class, `RouterTestingModule` can be imported instead of the `RouterModule` itself. 
+The routes that are relevant can be defined in the `RouterTestingModule`:
+    
+{% highlight coffeescript %}  
+    imports: [RouterTestingModule.withRoutes([/*List mock routes here*/])]
+{% endhighlight %}  
+
+If you don't feel like doing all that, you can also tell Angular to skip elements it doesn't know by adding `NO_ERRORS_SCHEMA` to the TestBed configuration:
+
+{% highlight coffeescript %}  
+    TestBed.configureTestingModule({
+        declarations: [ AppComponent ],
+        schemas: [ NO_ERRORS_SCHEMA ]
+    })
+{% endhighlight %}  
+
+To learn more about writing tests using Angular TestBed, I recommend reading this guide: [https://angular.io/guide/testing](https://angular.io/guide/testing).
+
+## 2. Calling the constructor
+
+A much simpler way to do unit testing is to simply call the constructor of the class you want to test.
+You should get an instance of each class that's needed in the constructor.
+Of course we want to mock these classes and as we saw in the Angular TestBed section, the Jasmine spy objects are the way to go.
+
+{% highlight coffeescript %}  
+    describe('AppComponent', () => {        
+        let component: AppComponent;
+
+        const mockCarBrandService = jasmine.createSpyObj('carBrandService', ['findAll']);
+        mockCarBrandService.findAll.and.returnValue(Observable.of([
+            { name: 'Mazda', country: 'Japan' },
+            { name: 'BMW', country: 'Germany' }
+        ]);
+
+        beforeEach(() => {
+            component = new AppComponent(mockCarBrandService);
+        });
+        ...
+    });
+{% endhighlight %}  
+
+Without the TestBed, you don’t have access to the view. 
+However, your tests will run much faster as there are less things to load. 
+When using TestBed, you’ll probably be including lots of dependencies just to make it work, giving you less control. 
+Something you do not want in unit testing. 
+Another difference with TestBed is that you have to call the lifecycle events yourself, again giving you more control over the code you’re testing.
+
+{% highlight coffeescript %}  
+    it('should find the car brand', () => {
+        component.ngOnInit();
+        const carBrand = component.getFavoriteCarBrand();
+        expect(carBrand.name).toEqual('Mazda');
+    }); 
+{% endhighlight %}  
+
+These kind of tests are much easier to write, give you more control, run faster, allow you to completely isolate the class and take away the hassle of getting the TestBed setup right. 
+Unless of course, you really want to test the view as well. 
+However, in my opinion, that’s not the point of unit testing. 
```

Comment:
Here you made the point that TestBed is not that ideal for unit testing. 
You could rephrase the title of the article to f.e.: "Testing in Angular 101" or something like that.

Reply:
Indeed, I think a lot of people (like me) first make the mistake to use TestBed for unit testing. I like the title, but I don't talk about protractor testing so maybe add Karma somewhere in it? "Testing in Angular with Karma 101" ?

Reply:
Sounds good, or "Unit and integration testing in Angular with Karma"?

Reply:
Hmm, but that would mean I have to elaborate more on integration testing :D

Reply:
Yeah, I agree not to elaborate too much on integration testing! And like you say, when you start writing about TestBed, tell them their stepping into the realm of integration tests, and give a short distinction between unit and integration tests.

Integration tests verify the interaction between multiple "components", and by components I don't necessarily mean Angular Components, but also the templates / services / other dependencies related to the "component" under test. 

I think it's perfectly valid to write integration tests with Protractor, but Angular now comes with an alternative tool where we don't have to start up the whole app in a (headless) browser to run our integration tests. And I'm pretty sure tests written in TestBed run much faster than protractor tests.

The next level is E2E tests where you need to verify your acceptance criteria. 
(Integration testing != E2E Testing)

So here you can verify your app can actually build & run. You should not be using TestBed for this, don't think you can, Protractor is much more powerful.

When people load up their main module in TestBed, I think they either don't split up their modules in feature modules, or they are trying to do an E2E Test scenario, where in that case, they should consider protractor or something similar.